pkg/api/pluginproxy/access_token_provider_test.go

```diff
@@ -91,4 +95,192 @@ func TestAccessToken(t *testing.T) {
+				So(accessToken, ShouldEqual, token["access_token"])
+
+				for _, v := range tokenCache.cache {
+					So(v.ExpiresOn.Unix(), ShouldEqual, expiresOn)
+				}
+			})
+
+			Convey("Should parse and cache access token with expires_on unixTimestamp as string field", func() {
+				expiresOn := time.Now().Add(6 * time.Hour).Unix()
+				token = map[string]interface{}{
+					"access_token":  "2YotnFZFEjr1zCsicMWpuu",
+					"token_type":    "example",
+					"expires_on":    strconv.FormatInt(expiresOn, 10),
+					"refresh_token": "tGzv3JOkF0XG5Qx2TlKWIA",
+				}
+				provider := newAccessTokenProvider(ds, pluginRoute)
+				accessToken, err := provider.getAccessToken(context.Background(), templateData)
+				So(err, ShouldBeNil)
+				So(accessToken, ShouldEqual, token["access_token"])
+
+				accessToken, err = provider.getAccessToken(context.Background(), templateData)
+				So(err, ShouldBeNil)
+				So(accessToken, ShouldEqual, token["access_token"])
+
+				for _, v := range tokenCache.cache {
+					So(v.ExpiresOn.Unix(), ShouldEqual, expiresOn)
+				}
+			})
+
+			Convey("Should parse and cache access token even if accesstoken has no expire_on or expire_in", func() {
+				token = map[string]interface{}{
+					"access_token":  "2YotnFZFEjr1zCsicMWpuu",
+					"token_type":    "example",
+					"refresh_token": "tGzv3JOkF0XG5Qx2TlKWIA",
+				}
+				provider := newAccessTokenProvider(ds, pluginRoute)
+				accessToken, err := provider.getAccessToken(context.Background(), templateData)
+				So(err, ShouldBeNil)
+				So(accessToken, ShouldEqual, token["access_token"])
+
+				accessToken, err = provider.getAccessToken(context.Background(), templateData)
+				So(err, ShouldBeNil)
+				So(accessToken, ShouldEqual, token["access_token"])
+
+				for _, v := range tokenCache.cache {
+					So(v.ExpiresOn.Unix(), ShouldAlmostEqual, time.Now().Unix())
+				}
+			})
+
+			Convey("Should prefer expire_on", func() {
+				expiresOn := time.Now().Add(6 * time.Hour).Unix()
+				expiresIn := 5200
+				token = map[string]interface{}{
+					"access_token":  "2YotnFZFEjr1zCsicMWpuu",
+					"token_type":    "example",
+					"expires_on":    strconv.FormatInt(expiresOn, 10),
+					"expires_in":    expiresIn,
+					"refresh_token": "tGzv3JOkF0XG5Qx2TlKWIA",
+				}
+				provider := newAccessTokenProvider(ds, pluginRoute)
+				t := time.Now()
+				accessToken, err := provider.getAccessToken(context.Background(), templateData)
+				So(err, ShouldBeNil)
+				So(accessToken, ShouldEqual, token["access_token"])
+
+				accessToken, err = provider.getAccessToken(context.Background(), templateData)
+				So(err, ShouldBeNil)
+				So(accessToken, ShouldEqual, token["access_token"])
+
+				for _, v := range tokenCache.cache {
+					So(v.ExpiresOn.Unix(), ShouldNotAlmostEqual, t.Unix()+int64(expiresIn))
+					So(v.ExpiresOn.Unix(), ShouldEqual, expiresOn)
+				}
+			})
+
+			Reset(func() {
+				tokenCache.cache = map[string]*jwtToken{}
+				token = map[string]interface{}{}
+			})
```

Comment:
Access to `tokenCache` should be mutually excluded therefore it should be surrounded by a call to `Lock` and `Unlock` similarly to [this](https://github.com/grafana/grafana/blob/f781cb58594ef989569c2726fb997fd1087b8e9c/pkg/models/datasource_cache_test.go#L235) example code.

In addition to this, in the future we want to [migrate away](https://github.com/grafana/grafana/tree/master/pkg#reduce-the-use-of-goconvey) from Convey therefore instead of using the Convey specific method `Reset`, I would suggest to declare a `clearTokenCache` function and call this at the beginning of it's test.